javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -12,31 +12,160 @@
 
 package java.util
 
+import java.lang.{reflect => jlr}
+
 import scala.scalajs.js
 
 import scala.annotation.tailrec
 
-import scala.reflect.ClassTag
-
 import ScalaOps._
 
 object Arrays {
 
+  /** Safe upcasting of an array, following run-time covariance. */
+  @inline private def arrayUpcast[A](a: Array[_ <: A]): Array[A] =
+    a.asInstanceOf[Array[A]]
+
+  @inline private def componentTypeOf[A](clazz: Class[_ <: Array[A]]): Class[_ <: A] =
+    clazz.getComponentType().asInstanceOf[Class[_ <: A]]
+
+  @inline private def componentTypeOf[A](a: Array[A]): Class[_ <: A] =
+    componentTypeOf(a.getClass())
+
+  /** A custom typeclass for the operations we need in `Arrays` to implement
+   *  the algorithms generically.
+   */
+  private sealed abstract class ArrayOps[A] {
```

Comment:
I do not understand why this class has to conflate so many things. Essentially, I see 3 concerns:

- Array operations on existing arrays (IIUC to abstract over array of primitive types)
- Array creation (to abstract over `Class` / templates + the quirks that come with the typings of `getComponentType()`)
- Element comparison.

This conflation shows in the implementation where there are plenty of `UnsupportedOperationExceptions` thrown.

From what I see here alone, I do not understand why this is necessary (as opposed to splitting the typeclass concerns).

Could you elaborate?

Reply:
Not sure how I had arrived at that design. It probably felt simpler while I was writing all this.

I pushed a commit that splits `ArrayOps` into 4 different components. This gets rid of all the `UnsupportedOperationException`s.